R/cv_estimate_rt.R

```diff
@@ -27,78 +27,83 @@
 #' y <- c(1, rpois(100, dnorm(1:100, 50, 15) * 500 + 1))
 #' cv <- cv_estimate_rt(y, korder = 3, nfold = 3, nsol = 30)
 #' cv
-cv_estimate_rt <- function(observed_counts,
-                           korder = 3L,
-                           dist_gamma = c(2.5, 2.5),
-                           nfold = 3,
-                           error_measure = c("mse", "mae", "deviance"),
-                           x = 1:n,
-                           lambda = NULL,
-                           maxiter = 1e6L,
-                           ...) {
+cv_estimate_rt <- function(
+    observed_counts,
+    korder = 3L,
+    dist_gamma = c(2.5, 2.5),
+    nfold = 3L,
+    error_measure = c("mse", "mae", "deviance"),
+    x = 1:n,
+    lambda = NULL,
+    maxiter = 1e6L,
+    delay_distn = NULL,
+    ...) {
 
   arg_is_pos_int(nfold)
   n <- length(observed_counts)
+  arg_is_length(n, x)
+  xin <- x
+  if (inherits(xin, "Date")) x <- as.numeric(x)
+  arg_is_numeric(x)
 
-  if (nfold == 1) cli::cli_abort("nfold must be greater than 1")
+  if (nfold == 1) cli_abort("nfold must be greater than 1")
 
   ## Run program one time to create lambda
   full_data_fit <- estimate_rt(
     observed_counts = observed_counts,
     korder = korder,
-    x = x,
+    x = xin,
     lambda = lambda,
+    maxiter = maxiter,
     ...)
 
-  ## Use values from the full data fit
   if (is.null(lambda)) lambda <- full_data_fit$lambda
-  weighted_past_counts <- full_data_fit$weighted_past_counts
-  x <- full_data_fit$x
 
-  # Cross validation
-  foldid = fold_calculator(n, nfold)
+  foldid <- c(0, rep_len(1:nfold, n - 2), 0)
```

Comment:
Can we randomly assign the indices?

Reply:
No. Because there's a structure to the data (a time series in this case), the splitting needs to reflect that structure. Evenly spaced splits (while ensuring that the first and last observation are always present) allows us to predict the held out set simply via interpolation.

Reply:
Right. Make sense. Thank you for the explanation!